### PR TITLE
Improving testing infrastructure

### PR DESCRIPTION
**Pull request by nhibberd**
ambiata/mundane#74

### CODE REVIEW

File: notion-core/src/main/scala/com/ambiata/notion/core/Key.scala [hunk 1]
```diff
@@ -50,4 +54,10 @@ object Key {
   def fromString(s: String): Option[Key] =
     Some(new Key(s.split("/").toVector.map(KeyName.unsafe)))
 
+  implicit def KeyOrder: Order[Key] =
+    Order.order((x, y) => x.name.?|?(y.name))
+
+  implicit def KeyOrdering =
```
Comment (markhibberd) [inline notion-core/src/main/scala/com/ambiata/notion/core/Key.scala]: Type annotation, `scala.Ordering`

File: notion-core/src/main/scala/com/ambiata/notion/core/S3OutputStream.scala [hunk 2]
```diff
@@ -11,16 +10,14 @@ import java.io._
 import scalaz._, Scalaz._, effect.IO
 
 object S3OutputStream {
-  def stream(address: S3Address, client: AmazonS3Client): RIO[OutputStream] = {
-    val tmpPath = uniqueFilePath
-    RIO.safe[OutputStream]({
-      val f = new BufferedOutputStream(new FileOutputStream(tmpPath.path))
+  def stream(address: S3Address, client: AmazonS3Client): RIO[OutputStream] = for {
+    t <- LocalTemporary.random.file
```
Comment (markhibberd) [inline notion-core/src/main/scala/com/ambiata/notion/core/S3OutputStream.scala]: Not sure about this one, given that there is a call to unsafe performIO and this is delaying when the file would be deleted, I am not sure this is really the same code, maybe it is ok, I would just want to be sure nothing would be leaked if a long running process was calling this a lot.

File: notion-core/src/main/scala/com/ambiata/notion/core/StoreTemporary.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import scalaz.{Store => _}
+
+case class StoreTemporary(t: T, path: String) {
+  val client = Clients.s3
+
+  def store: RIO[Store[RIO]] =
+    if (t == T.Posix)     posixStore
+    else if (t == T.S3)   s3Store
+    else if (t == T.Hdfs) hdfsStore
+    else                  RIO.fail("Could not create store. Unsupported temporary type")
```
Comment (markhibberd) [inline notion-core/src/main/scala/com/ambiata/notion/core/StoreTemporary.scala]: Can't you pattern match on `t` this isn't possible and the compiler can check that right?

File: notion-core/src/main/scala/com/ambiata/notion/core/StoreTemporary.scala [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import com.ambiata.saws.core.Clients
+import com.ambiata.saws.s3._
+import org.apache.hadoop.conf.Configuration
+
+import scalaz._, Scalaz._
+
+import scalaz.{Store => _}
+
+case class StoreTemporary(t: T, path: String) {
+  val client = Clients.s3
```
Comment (markhibberd) [inline notion-core/src/main/scala/com/ambiata/notion/core/StoreTemporary.scala]: Not sure this should be using a hard coded client.

File: notion-core/src/test/scala/com/ambiata/notion/core/LocationIOSpec.scala [hunk 5]
```diff
@@ -29,104 +28,75 @@ class LocationIOSpec extends Specification with ForeachTemporaryType with ScalaC
...
-    withConf(conf => f(LocationIO(conf, Clients.s3)))
-}
+  def list = prop((loc: LocationTemporary, dp: DistinctPair[Ident], data: String) => for {
+    p <- loc.location
+    i <- loc.io
+    _ <- i.writeUtf8(p </> FilePath.unsafe(dp.first.value), data)
+    _ <- i.writeUtf8(p </> FilePath.unsafe(dp.second.value), data)
+    r <- i.list(p)
+    l = r.length
+  } yield r.map(_.render.split("/").last).toSet -> l ==== Set(dp.first.value, dp.second.value) -> 2)
```
Comment (markhibberd) [inline notion-core/src/test/scala/com/ambiata/notion/core/LocationIOSpec.scala]: :+1:

File: notion-core/src/test/scala/com/ambiata/notion/core/LocationTemporary.scala [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import scalaz._, Scalaz._
+
+case class LocationTemporary(t: TemporaryType, path: String) {
+  val client = Clients.s3
+
+  def location: RIO[Location] =
+    if (t == TT.Posix)     localLocation
+    else if (t == TT.S3)   s3Location
+    else if (t == TT.Hdfs) hdfsLocation
+    else                   RIO.fail("Could not create location. Unsupported temporary type")
```
Comment (markhibberd) [inline notion-core/src/test/scala/com/ambiata/notion/core/LocationTemporary.scala]: Same as above, pattern match instead of if.

Comment (markhibberd) [inline notion-core/src/test/scala/com/ambiata/notion/core/LocationTemporary.scala]: Or you could fold if TT had one.

**Comment by markhibberd** [comment]
Couple of questions inline, but looks good.

File: notion-core/src/main/scala/com/ambiata/notion/core/S3OutputStream.scala (hunk 2 above)
Comment (nhibberd) [inline notion-core/src/main/scala/com/ambiata/notion/core/S3OutputStream.scala]: As discussed, is the same

File: notion-core/src/main/scala/com/ambiata/notion/core/S3Store.scala [hunk 7]
```diff
@@ -221,15 +221,8 @@ case class S3Store(s3: S3Prefix, client: AmazonS3Client, cache: DirPath) extends
     def withInputStream(key: Key)(f: InputStream => RIO[Unit]): RIO[Unit] =
       readOnly.unsafe.withInputStream(key)(f)
 
-    def withOutputStream(key: Key)(f: OutputStream => RIO[Unit]): RIO[Unit] = {
-      val unique = Key.unsafe(UUID.randomUUID.toString)
-      local.unsafe.withOutputStream(unique)(f) >> local.unsafe.withInputStream(unique)(in =>
-        run { (s3 | key.name).putStreamWithMetadata(in, S3Address.ReadLimitDefault, {
-          val metadata = S3.ServerSideEncryption
-          metadata.setContentLength((local.root </> keyToFilePath(unique)).toFile.length)
```
Comment (charleso) [inline notion-core/src/main/scala/com/ambiata/notion/core/S3Store.scala]: I'm assuming removing this was intentional?

Comment (nhibberd) [inline notion-core/src/main/scala/com/ambiata/notion/core/S3Store.scala]: Replacing it? with :225 S3OutputStream.stream

Comment (charleso) [inline notion-core/src/main/scala/com/ambiata/notion/core/S3Store.scala]: > Replacing it? 

Does `stream` use SSE? I guess we don't need the contentLength any more.

Comment (nhibberd) [inline notion-core/src/main/scala/com/ambiata/notion/core/S3Store.scala]: Uses the PutObjectRequest in the sdk. Puts file rather than writes stream

**Pull request merged**